Improve execution name readability

wayner0628 opened the pull request:
## Tracking issue
<!--
If your PR fixes an open issue, use `Closes #999` to link your PR with the issue.
Example: Closes #999

If your PR is related to an issue or PR, use `Related to #999` to link your PR.
Example: Related to #999

Remove this section if not applicable
-->

## Why are the changes needed?
- Improve execution name readability

<!--
Please clarify why the changes are needed. For instance,
1. If you propose a new API, clarify the use case for a new API.
2. If you fix a bug, you can clarify why it is a bug.
-->

## What changes were proposed in this pull request?
- Add `EnableHumanHash` config in [flyteadmin/pkg/runtime/interfaces/application_configuration.go](https://github.com/flyteorg/flyte/pull/5637/commits/5f37b11138bdd32231151d40d794d30a0986b41f#diff-e4e11bcd1e36c8870c82a37fcab52de08d6de4911b1fcf83195cea4156f9a0bb)
- Add human hash feature in [flyteadmin/pkg/common/executions.go](https://github.com/flyteorg/flyte/pull/5637/commits/815f597d4f66aa56d25c8ce89fd482171e7c570c#diff-331f07e355b610d0581bd6dcc08b8f31bddcea598ecafeaab61836aa7b7c77bc)

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
2. If there is design documentation, please add the link.
-->

## How was this patch tested?
unit testing and integration testing
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

### Setup process
To enable human-readable execution names, add the following configuration:

```yaml
flyteadmin:
  featureGates:
    enableHumanHash: true
```

### Screenshots
![Screenshot 2024-08-11 at 9 55 04 PM](https://github.com/user-attachments/assets/1702eaf9-73d7-415a-8f1e-0aa25907d433)

## Check all the applicable boxes <!-- Follow the above conventions to check the box -->

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/2678
<!-- Add related pull requests for reviewers to check -->

## Docs link

<!-- Add documentation link built by CI jobs here, and specify the changed place -->

pingsutw commented on flytectl/cmd/create/execution_util.go:
```diff
@@ -150,7 +153,12 @@ func recoverExecution(ctx context.Context, executionName string, project string,
 
 	if len(targetExecName) == 0 {
-		targetExecName = "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19]
+		config := runtime.NewApplicationConfigurationProvider()
```
Let's update flyteadmin first. Ideally, the ID should only be generated from the backend instead of the client.

pingsutw commented on flyteadmin/pkg/common/executions.go:
```diff
@@ -3,17 +3,24 @@ package common
 	rand.Seed(seed)
+	if enableHumanHash {
+		hashKey := []byte(rand.String(ExecutionIDLength))
+		result, _ := humanhash.Humanize(hashKey, 3)
```
should we handle the error here?
```suggestion
		result, err := humanhash.Humanize(hashKey, 3)
```

pingsutw commented on flyteadmin/pkg/async/schedule/aws/workflow_executor.go:
```diff
@@ -206,8 +215,10 @@ func (e *workflowExecutor) run() error {
-
+		executionRequest, err := e.formulateExecutionCreateRequest(launchPlan, scheduledWorkflowExecutionRequest.KickoffTime)
+		if err != nil {
+			continue
```
We should log an error here

eapolinario commented on flyteadmin/pkg/manager/impl/util/shared.go:
```diff
@@ -15,17 +15,23 @@ import (
+		return request.Name, nil
 	}
-	return common.GetExecutionName(time.Now().UnixNano())
+	config := runtime.NewApplicationConfigurationProvider()
+	executionName, err := common.GetExecutionName(time.Now().UnixNano(), config.GetTopLevelConfig().FeatureGates.EnableHumanHash)
+	if err != nil {
+		return "", err
+	}
+	return executionName, nil
```
we shouldn't need this once we go back to the original signature for `GetExecutionName`.

eapolinario commented on flyteadmin/pkg/runtime/interfaces/application_configuration.go:
```diff
@@ -50,6 +50,7 @@ type PostgresConfig struct {
 
 type FeatureGates struct {
 	EnableArtifacts bool `json:"enableArtifacts" pflag:",Enable artifacts feature."`
+	EnableHumanHash bool `json:"enableHumanHash" pflag:",Enable humanhash feature."`
```
Can we rename this to `EnableFriendlyNames`?

```suggestion
	EnableFriendlyNames bool `json:"enableFriendlyNames" pflag:",Enable generation of friendly execution names feature."`
```

eapolinario commented on flyteadmin/scheduler/executor/executor_impl.go:
```diff
@@ -66,10 +68,22 @@ func (w *executor) Execute(ctx context.Context, scheduledTime time.Time, s model
 		return err
 	}
 
+	executionName := ""
+	config := runtime.NewApplicationConfigurationProvider()
+	if config.GetTopLevelConfig().FeatureGates.EnableHumanHash {
+		executionName, err = common.GetExecutionName(time.Now().UnixNano(), config.GetTopLevelConfig().FeatureGates.EnableHumanHash)
+		if err != nil {
+			logger.Errorf(ctx, "failed to generate execution name for schedule %+v due to %v", s, err)
+			return err
+		}
+	} else {
+		executionName = "f" + strings.ReplaceAll(executionIdentifier.String(), "-", "")[:19]
+	}
```
Are friendly names important for scheduled executions?

eapolinario commented on flyteadmin/pkg/common/executions.go:
```diff
@@ -1,20 +1,33 @@
-	return fmt.Sprintf(ExecutionStringFormat, rand.String(ExecutionIDLength-1))
+	if enableHumanHash {
+		hashKey := []byte(rand.String(ExecutionIDLength))
+		result, err := humanhash.Humanize(hashKey, 3)
```
IIUC, by setting this to 3 we get about 16M unique ids to choose from. The reason is that humanhash maintains a list of [unique 256 words](https://github.com/wolfeidau/humanhash/blob/master/human_hash.go#L28-L64), so 256^3 = 16.8M. 

We should have a conversation about whether this is enough for most big deployments. My gut feeling is that it's not sufficient. How about we extend this to 4? If we do this and end up choosing the largest word in that set 4 times we're still under the 63, which is the max char limit for an execution id. Out of curiosity, the largest word in that dataset is "massachusetts" at 13 characters, so that would be 55 characters (13*4 + 3 to account for the separators).

wayner0628 commented:
Hello @pingsutw @eapolinario,

I've made the following changes based on your feedback:

1. Removed all unnecessary error handlers.
2. Added a comment explaining why errors in humanhash generation can be safely ignored.
3. Updated the humanhash package version to `v1.1.0`.
4. Renamed the configuration setting to `EnableFriendlyNames`.
5. Extend the name to `4` subwords.
6. Now reading the config directly within the `getExecutionName` function, and adjusted the file path accordingly.
7. Updated the unit tests to use mocking objects, as the config is now read within the function.

Please review these changes and let me know if you have any suggestions or concerns.

Thank you!

wayner0628 commented:
Hi @pingsutw @eapolinario,

I've thoroughly investigated the issue but couldn't find any connection between my changes and the CI unit test failure. Additionally, there are no error messages logged that could provide further insight. Could you please help retrigger the CI build? Thank you!

pingsutw reviewed (approved):
I just reverted the flytectl changes. It seems like we can't change that in the same PR. You could create a follow-up PR to update flytectl.

wayner0628 commented:
> I just reverted the flytectl changes. It seems like we can't change that in the same PR. You could create a follow-up PR to update flytectl.

I noticed some issues still persist, so I'll continue checking them.

pingsutw commented:
@ddl-ebrown, we could make `suffix` opt-in. Need to add one more field to the config map. something like
```
flyteadmin
  featureGates:
    enableFriendlyNames: true
  execution_name_with_random_suffix: true
```

Feel free to open a PR!

kumare3 commented:
This is pretty cool - cc @katrogan
